Src/ST-LIB_LOW/StateMachine/StateMachine.cpp

```diff
@@ -269,11 +280,20 @@ void StateMachine::force_change_state(uint8_t new_state) {
 
 	unregister_all_timed_actions(current_state);
 	exit_state(current_state);
-
+	#ifdef SIM_ON
+		old_state=current_state;
+	#endif
 	current_state = new_state;
 
 	enter_state(current_state);
 	register_all_timed_actions(current_state);
+
+	#ifdef SIM_ON
+		for(uint8_t i=1; i<=SharedMemory::state_machine_sm[0]; i++){
+			if(SharedMemory::state_machine_sm[i] == old_state)
+				SharedMemory::update_current_state(i,current_state);
+		}
```

Comment:
```suggestion
        SharedMemory::update_current_state(the_counter_you_stored_in_the_constructor, current_state);
```
Simplifies a lot the logic

Reply:
yes, this is it.